sov-modules/sov-modules-impl/prover-incentives/src/call.rs

```
@@ -0,0 +1,170 @@
+    ) -> Result<sov_modules_api::CallResponse> {
+        let mut response = CallResponse::default();
+
+        // Get the prover's old balance.
+        if let Some(old_balance) = self.bonded_provers.get(context.sender(), working_set) {
+            // Transfer the bond amount from the sender to the module's address.
+            // On failure, no state is changed
+            let coins = Coins {
+                token_address: self
+                    .bonding_token_address
+                    .get(working_set)
+                    .expect("Bonding token address must be set"),
+                amount: old_balance,
+            };
+            // Try to unbond the entire balance
+            // If the unbonding fails, no state is changed
+            self.bank
+                .transfer_from(&self.address, context.sender(), coins, working_set)?;
+
+            // Update our internal tracking of the total bonded amount for the sender.
+            self.bonded_provers.set(context.sender(), 0, working_set);
+
+            // Emit the unbonding event
+            response.add_event(
+                "unbonded_prover",
+                &format!("amount_withdrawn: {old_balance:?}"),
+            );
+        }
+
+        Ok(response)
+    }
+
+    /// Try to process a zk proof, if the prover is bonded.
+    pub(crate) fn process_proof(
+        &self,
+        proof: &[u8],
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        let mut response = CallResponse::default();
+
+        // Get the prover's old balance.
+        if let Some(old_balance) = self.bonded_provers.get(context.sender(), working_set) {
+            // Check that the prover has enough balance to process the proof.
+
+            let minimum_bond = self
+                .minimum_bond
+                .get(working_set)
+                .expect("Minimum bond must have been set at genesis");
```

Is it intended to panic, instead of returning result?

Generally I don't see the "error" return in this method. Why is it "Result" then?

---

Updated in latest iteration